include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -207,7 +261,73 @@ struct transform_init
+        if (__items_to_transform == 32)
+        {
+            typename _AccLocal::value_type __res = __unary_op(__adjusted_global_id, __acc...);
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 1, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 2, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 3, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 4, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 5, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 6, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 7, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 8, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 9, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 10, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 11, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 12, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 13, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 14, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 15, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 16, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 17, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 18, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 19, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 20, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 21, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 22, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 23, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 24, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 25, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 26, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 27, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 28, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 29, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 30, __acc...));
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + 31, __acc...));
```

Comment:
You can reduce the amount of repeated code by using recursive invocations of a class template. Something like:

```c++
template<typename T>
struct _Unroll;

template<int I, int... Is>
struct _Unroll<std::integer_sequence<int, I, Is...>>
{
    template<typename _R, typename _BinOp, typename _UnOp, typename... _Acc>
    static void apply(_R& __res, _BinOp&& __binary_op, _UnOp&& __unary_op, ::std::size_t __adjusted_global_id, _Acc&&... __acc)
    {
        __res = __binary_op(__res, __unary_op(__adjusted_global_id + I, __acc...));
        _Unroll<std::integer_sequence<int, Is...>>::apply(__res, __binary_op, __unary_op, __adjusted_global_id, __acc...);
    }
};

template<>
struct _Unroll<std::integer_sequence<int>>
{
    template<typename _R, typename _BinOp, typename _UnOp, typename... _Acc>
    static void apply(_R& __res, _BinOp&& __binary_op, _UnOp&& _unary_op, ::std::size_t __adjusted_global_id, _Acc&&... __acc)
    {
    }
};

int main()
{
    int x = 0;
    std::vector<int> v(32, 5);
    _Unroll<std::make_integer_sequence<int, 32>>::apply(
        x,
        [](int x, int y) { return x + y; },
        [](std::size_t i, const std::vector<int>& vec) { return vec[i]*2; },
        0,
        v
    );
}
```
There's probably a more modern way to do it with fold expressions, but I can't get the syntax to play nice.

Reply:
Thanks, Adam. The updated version seems to achieve a similar effect by using an unroll compiler directive.